include/ignition/gazebo/detail/ComponentStorage.hh

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_GAZEBO_DETAIL_COMPONENTSTORAGE_HH_
+#define IGNITION_GAZEBO_DETAIL_COMPONENTSTORAGE_HH_
+
+#include <cstddef>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+#include "ignition/gazebo/Entity.hh"
+#include "ignition/gazebo/Types.hh"
+#include "ignition/gazebo/components/Component.hh"
+#include "ignition/gazebo/config.hh"
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace detail
+{
+/// \brief Enum class that lets external users of ComponentStorage know what
+/// type of action took place when a component was added to an entity.
+enum class ComponentAdditionResult
+{
+  /// \brief The type of the component added to an entity is a type that wasn't
+  /// previously associated with the entity.
+  NEW_ADDITION = 0,
+  /// \brief The type of the component added to an entity was a previously
+  /// existing component type for that entity.
+  RE_ADDITION = 1,
+  /// \brief The component was not added to the entity because the entity
+  /// already has a valid instance of this component type. Instead, the data for
+  /// the valid instance should be modified with the new component data.
+  MODIFICATION = 2,
+  /// \brief The attempt to add the component to the entity failed, probably
+  /// because the entity does not exist.
+  FAILED_ADDITION = 3
+};
+
+class IGNITION_GAZEBO_HIDDEN ComponentStorage
```

Comment:
Documentation has been added in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/52f8c7eedbd44629b97860dcf804207d49414788

> Also, since it serves a different purpose from the old `ComponentStorage`

This class actually doesn't serve a different purpose from the old `ComponentStorage` - both this class and the old storage class are meant to store components that have been created. The main difference between the two classes is that the new class does not store components sequentially in memory, while the old class does. This change/difference allows for the new class to keep an internal mapping of which component an entity belongs to, which was not possible with the old class (see https://github.com/ignitionrobotics/ign-gazebo/pull/856#discussion_r659100121 for more information)

Reply:
Since we removed this class in #943, I'm marking this as resolved